language/bytecode-verifier/src/abstract_state.rs

```diff
@@ -129,436 +110,178 @@ impl AbstractState {
         self.locals.insert(idx, value);
     }
 
-    /// Return true if self may safely be destroyed
-    pub fn is_safe_to_destroy(&self) -> bool {
-        self.locals.values().all(|x| x.is_safe_to_destroy())
-            && self.globals.values().all(|x| x.is_empty())
+    /// checks if a local may be safely destroyed
+    pub fn is_local_safe_to_destroy(&self, idx: LocalIndex) -> bool {
+        match self.locals[&idx] {
+            AbstractValue::Reference(_) => true,
+            AbstractValue::Value(Kind::All) | AbstractValue::Value(Kind::Resource) => false,
+            AbstractValue::Value(Kind::Unrestricted) => !self.is_local_borrowed(idx),
+        }
     }
 
-    /// Return true if local@idx may safely be destroyed
-    pub fn is_local_safe_to_destroy(&self, idx: LocalIndex) -> bool {
-        self.local(idx).is_safe_to_destroy()
+    /// checks if the stack frame of the function being analyzed can be safely destroyed
+    pub fn is_frame_safe_to_destroy(&self) -> bool {
+        self.locals.values().all(|x| x.is_unrestricted_value())
+            && !self.is_nonce_borrowed(self.frame_root)
     }
 
     /// destroys local@idx
-    /// call only if self.is_local_safe_to_destroy(idx) returns true
     pub fn destroy_local(&mut self, idx: LocalIndex) {
+        checked_verify!(self.is_local_safe_to_destroy(idx));
         let local = self.locals.remove(&idx).unwrap();
         match local {
             AbstractValue::Reference(nonce) => self.destroy_nonce(nonce),
-            AbstractValue::Value(kind, borrowed_nonces) => {
-                checked_verify!(kind == Kind::Unrestricted && borrowed_nonces.is_empty());
+            AbstractValue::Value(kind) => {
+                checked_verify!(kind == Kind::Unrestricted);
             }
         }
     }
 
-    /// nonce must be fresh
+    /// adds nonce to borrow graph
+    /// nonce must be freshly allocated
     pub fn add_nonce(&mut self, nonce: Nonce) {
-        self.partition.add_nonce(nonce);
+        self.borrow_graph.add_nonce(nonce);
     }
 
-    /// destroys nonce
-    /// borrows of nonce become borrows of any nonce' such that nonce borrows from nonce'
+    /// destroys nonce from borrow graph
```

Comment:
One does not destroy something "from". Perhaps you mean to say "remove nonce from borrow graph and destroy it". Is it possible to say that in a post condition?

Reply:
I have renamed destroy_nonce to remove_nonce and also changed the documentation.  I have  added comprehensive specifications to borrow_graph.rs.